wallet, refactor: #24584 follow-ups

josibake opened the pull request:
This PR is to address follow-ups for #24584, specifically:

* Remove redundant, hard-to-read code by adding a new `OutputType` and adding shuffle, erase, and push_back methods for `CoinsResult`
* Add missing `BOOST_ASSERT` to unit test
* Ensure functional test only runs if using descriptor wallets
* Improve readability of `AttemptSelection` by removing triple-nested if statement

Note for reviewers: commit `refactor: add new helper methods` should throw an "unused function warning"; the function is used in the next commit. Also, commit `wallet: switch to new shuffle, erase, push_back` will fail to compile, but this is fixed in the next commit with a scripted-diff. the commits are separate like this (code change then scripted-diff) to improve legibility.

josibake committed:
refactor: add UNKNOWN OutputType

add to enum, array and handle UNKNOWN in various case statements

josibake committed:
refactor: add new helper methods

add Shuffle, Erase, and Add to CoinsResult struct
add a helper function for mapping TxoutType to OutputType

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

josibake committed:
wallet: switch to new shuffle, erase, push_back

switch to new methods, remove old code. this also
updates the Size, All, and Clear methods to now use
the coins map.

this commit is not strictly a refactor because previously
coin selection was never run over the UNKNOWN type until the last
step when being run over all. now that we are iterating over each,
it is run over UNKNOWN but this is expected to be empty most of the time.

Co-authored-by: furszy <matiasfurszyfer@protonmail.com>

josibake committed:
test: add missing BOOST_ASSERT

this was missed in the original PR

josibake committed:
test: only run test for descriptor wallets

since this test uses bech32m, we skip unless sqlite is used, which is the
same as checking if we are using descriptor wallets or not

josibake committed:
refactor: improve readability for AttemptSelection

it was pointed out by a few reviewers that the code block at the end
of attempt selection was difficult to follow and lacked comments.

refactor to get rid of triple nested if statement and improve
readibility.

josibake commented:
tagging @furszy , @LarryRuane , @Xekyo , and @aureleoules for review as this addresses feedback from y'all on #24584 . thanks for the thorough review and great suggestions, let me know if i missed anything!

aureleoules commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -124,6 +124,7 @@ def set_test_params(self):
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
+        self.skip_if_no_sqlite()
```
this doesn't seem to fix it :/
```
python test/functional/wallet_avoid_mixing_output_types.py
...
test_framework.authproxy.JSONRPCException: Legacy wallets cannot provide bech32m addresses (-8)
```

josibake replied:
i think this is the best we can do right now with the test framework. if no sqlite, skip. if sqlite and bdb, pass the `--descriptors` flag, else the test will fail

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -79,30 +79,70 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *walle
+{
+    for (auto& it : coins) {
+        auto& vec = it.second;
+        auto i = find_if(vec.begin(), vec.end(), [&](const COutput &c) { return preset_coins.count(c.outpoint);});
```
```suggestion
        auto i = std::find_if(vec.begin(), vec.end(), [&](const COutput &c) { return preset_coins.count(c.outpoint);});
```

aureleoules reviewed (commented):
Concept ACK
This simplifies the code and is easier to read!
I left some comments.

w0xlt reviewed (commented):
Approach ACK.
These changes improve code readability and organization.

furszy commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1966,6 +1966,10 @@ bool DescriptorScriptPubKeyMan::SetupDescriptorGeneration(const CExtKey& master_
         desc_prefix = "tr(" + xpub  + "/86'";
         break;
     }
+    case OutputType::UNKNOWN: {
+        // Do nothing if the address is of UNKNOWN OutputType
+        return false;
+    }
```
Could assert(false) here.
`DescriptorScriptPubKeyMan::SetupDescriptorGeneration` is called from `CWallet::SetupDescriptorScriptPubKeyMans` and this latter function is already skipping this type.

josibake replied:
good point, `SetupDescriptorGeneration` is actually defined on the `DescriptorScriptPubKeyMan` and we should never have a spk_m setup for the unknown type, so assert(false) is better than just returning a bool.

furszy reviewed (commented):
3e6cbb2d doesn't compile without ac0585d4. Better to squash them.

furszy reviewed (commented):
Another point, `CanGetAddresses` and `GetActiveScriptPubKeyMans` loop through the available output types vector as well. And they are not included in c81cb2aa

josibake commented:
> Another point, `CanGetAddresses` and `GetActiveScriptPubKeyMans` loop through the available output types vector as well. And they are not included in [c81cb2a](https://github.com/bitcoin/bitcoin/commit/c81cb2aa86069e916ec2a50f2d04cd666f56d2c0)

both functions you mentioned call `GetScriptPubKeyMan()`, which checks if there is a spk_m for the specified output type in either `m_internal_spk_managers` or `m_external_spk_managers`. spk_ms are setup by `SetupDescriptorPubKeyMans`, where we are skipping the unknown type

josibake commented on src/wallet/spend.h:
```diff
@@ -38,6 +38,7 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* walle
  * the CoinsResult struct as if it was a vector
  */
 struct CoinsResult {
+    std::map<OutputType, std::vector<COutput>> coins;
```
id prefer to keep it simple for now. im not convinced it's worth re-writing the `coinselection_tests` to create TxoutTypes just to map them back to COutPuts and to add methods just for checking the size in the `availablecoins_tests`

josibake commented:
> [3e6cbb2](https://github.com/bitcoin/bitcoin/commit/3e6cbb2d4aa8452f73e43a62d5b4168d212745bf) doesn't compile without [ac0585d](https://github.com/bitcoin/bitcoin/commit/ac0585d4dd0174edd841ac3fc9d0799a6619a804). Better to squash them.

I split the changes into two commits to make the review easier and to be able to do the second part as a scripted-diff. This is mentioned in the commit message and the description, so it shouldn't be a surprise to reviewers

josibake commented:
force pushed to address comments from @furszy and @aureleoules ; `git range-diff master e40eaf6 7089df5`

furszy commented:
> > [3e6cbb2](https://github.com/bitcoin/bitcoin/commit/3e6cbb2d4aa8452f73e43a62d5b4168d212745bf) doesn't compile without [ac0585d](https://github.com/bitcoin/bitcoin/commit/ac0585d4dd0174edd841ac3fc9d0799a6619a804). Better to squash them.
> 
> I split the changes into two commits to make the review easier and to be able to do the second part as a scripted-diff. This is mentioned in the commit message and the description, so it shouldn't be a surprise to reviewers

Isn't about being a surprise or not. It's about being able to compile each commit separately while traversing the repository history. It's useful when you are tracking bugs to know where test started failing.
Plus AFAIK, it's a repository "policy".

adam2k commented:
Approach ACK.  I'm leaving a few minor nits.  Feel free to ignore or take them into consideration.

adam2k commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,49 @@ void CoinsResult::clear()
+static OutputType getOutputType(TxoutType type, bool is_from_p2sh)
+{
+    switch (type) {
+        case TxoutType::WITNESS_V1_TAPROOT:
+            return OutputType::BECH32M;
+        case TxoutType::WITNESS_V0_KEYHASH:
+        case TxoutType::WITNESS_V0_SCRIPTHASH:
+            if (is_from_p2sh) return OutputType::P2SH_SEGWIT;
+            else return OutputType::BECH32;
+            break;
+        case TxoutType::SCRIPTHASH:
+        case TxoutType::PUBKEYHASH:
+            return OutputType::LEGACY;
+            break;
```
nit: two of the `case` statements with the early `return` have `break;` and the first one does not. Could this be more consistent without the `break;` statements?

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,49 @@ void CoinsResult::clear()
+    coins[type].emplace_back(out);
+}
+
+static OutputType getOutputType(TxoutType type, bool is_from_p2sh)
```
In a0abea26692e63e9205ec68bdbc04cfad8c95a3d "refactor: add new helper methods"

nit: We prefer UpperCamelCase for function names.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,49 @@ void CoinsResult::clear()
+    }
+}
+
+void CoinsResult::push_back(OutputType type, const COutput& out)
```
In a0abea26692e63e9205ec68bdbc04cfad8c95a3d "refactor: add new helper methods"

nit: We prefer UpperCamelCase for function names.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,49 @@ void CoinsResult::clear()
+    }
+}
+
+void CoinsResult::shuffle(FastRandomContext& rng_fast)
```
In a0abea26692e63e9205ec68bdbc04cfad8c95a3d "refactor: add new helper methods"

nit: We prefer UpperCamelCase for function names.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -105,6 +105,49 @@ void CoinsResult::clear()
     other.clear();
 }
 
+void CoinsResult::erase(std::set<COutPoint>& preset_coins)
```
In a0abea26692e63e9205ec68bdbc04cfad8c95a3d "refactor: add new helper methods"

nit: We prefer UpperCamelCase for function names.

achow101 reviewed (commented):
41703e57cc0d2693533240ed916b9e55283433c6 does not compile

josibake committed:
scripted-diff: Uppercase function names

Change `CoinsResult` functions to uppercase to be consistent with
the style guide.

-BEGIN VERIFY SCRIPT-
git grep -l "available_coins" | grep -v mempool_stress.cpp | xargs sed -i "s/available_coins\.\(size\|all\|clear\)/available_coins\.\u\1/"
git grep -l AvailableCoins | xargs sed -i "/AvailableCoins/ s/\(all()\|size()\|clear()\)/\u\1/"
sed -i "s/\(clear()\|all()\|size()\)/\u&/g" src/wallet/spend.h
sed -i "/CoinsResult::/ s/\(clear()\|all()\|size()\)/\u&/" src/wallet/spend.cpp
sed -i "s/result.size/result.Size/" src/wallet/spend.cpp
sed -i "s/this->size/this->Size/" src/wallet/spend.cpp
-END VERIFY SCRIPT-

josibake commented:
> > > [3e6cbb2](https://github.com/bitcoin/bitcoin/commit/3e6cbb2d4aa8452f73e43a62d5b4168d212745bf) doesn't compile without [ac0585d](https://github.com/bitcoin/bitcoin/commit/ac0585d4dd0174edd841ac3fc9d0799a6619a804). Better to squash them.
> > 
> > 
> > I split the changes into two commits to make the review easier and to be able to do the second part as a scripted-diff. This is mentioned in the commit message and the description, so it shouldn't be a surprise to reviewers
> 
> Isn't about being surprised by it or not. It's about being able to compile each commit separately while traversing the repository history. It's useful when you are tracking bugs to know where test started failing. Plus AFAIK, it's a repository "policy".

Good point re: repository history; I squashed the two together so it compiles now

josibake commented:
force pushed to rebase, address feedback + a few minor fixups, and added a scripted-diff to uppercase the `CoinsResult` functions. `git range-diff master 7089df5 8cd21bb`

aureleoules reviewed (commented):
ACK 8cd21bb2799d37ed00dc9d0490bb5f5f1375932b.
Verified all the comments of the original PR have been addressed.
This PR makes `available_coins` easier to use and simplifies the code overall.

LarryRuane commented on src/wallet/spend.h:
```diff
@@ -34,26 +34,22 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* walle
-
-    /** Other is a catch-all for anything that doesn't match the known OutputTypes */
-    std::vector<COutput> other;
+    std::map<OutputType, std::vector<COutput>> coins;
```
Perhaps a `std::map` is overkill since there are only a few fixed types, and they're represented as small integers (since it's an `enum`). It seems like an array, with type as the array index, would be more appropriate. I started to leave a suggestion here to make this an array, but first wanted to check that suggestion would work, so I made the necessary changes and pushed it up to a branch: https://github.com/LarryRuane/bitcoin/commits/suggestions-25734

That branch directly modifies only the second commit "refactor: add new helper methods", then I cherry-picked the following commits to work with that commit (so the changes to those commits are only side-effects).

I like the way this came out, but I understand if you would rather not lose any ACKs or take the trouble to cherry-pick this branch. Just a suggestion.

josibake replied:
Thanks for the review and suggestions, @LarryRuane and sorry for the slow response. I do like this approach. Will definitely keep this in mind if I am working on the `CoinsResult` struct again soon

LarryRuane reviewed (commented):
Concept, code review ACK 8cd21bb2799d37ed00dc9d0490bb5f5f1375932b

Summary of my review: I made a [branch](https://github.com/LarryRuane/bitcoin/commits/suggestions-25734) to show suggestions, feel free to cherry-pick.

furszy commented on src/wallet/spend.h:
```diff
@@ -34,26 +34,18 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* walle
  *
- * Clear(), Size() methods are implemented so that one can interact with
- * the CoinsResult struct as if it was a vector
+ * Size(), Clear(), Erase(), Shuffle(), and Add() methods are implemented to
+ * allow easy interaction with the struct.
```
I think that this comment is more suited for a commit description, not for the sources. It doesn't provide any meaningful information.

josibake replied:
good point, I'll remove if I touch this in a later PR

furszy reviewed (approved):
utACK 8cd21bb2. Left a small, non-blocking, comment.

Plus, I would also like to have the `GetCoinsByType` that @LarryRuane suggested (would remove the `ByType` wording as the type filter is explicitly provided as a method argument).

But, overall, it's a non-blocking comment. We can move forward with it.

jonatack commented on src/wallet/spend.cpp:
```diff
@@ -79,30 +79,68 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *walle
+void CoinsResult::Add(OutputType type, const COutput& out)
+{
+    coins[type].emplace_back(out);
+}
```
Quick look on GitHub while waiting for a build:
 - any reason not to inline some of these methods? i.e. remove it from the .cpp and change the .h file to the following:
```cpp
void Add(OutputType type, const COutput& output) { coins[type].emplace_back(out); }
```
  same for Clear() (and maybe Shuffle() if it's a hotspot)

- naming nit: a param named `out` reads like an out-param; maybe `output` instead

josibake replied:
Thanks for the review @jonatack , great suggestions. Generally speaking, is there a heuristic we use in core for when to inline vs not?

Also, will include your suggestions if working with `CoinsResult` again

jonatack replied:
> when to inline vs not?

Inlining improves performance with the potential trade-off of increasing the code in the .h file and making changes to it slower to compile, so I think about it as: inline where performance is important, or for methods that are both very short (e.g. a single line or maybe 2-3) and that don't add an include dependency to the header file (rather than in the implementation file).

achow101 commented:
ACK 8cd21bb2799d37ed00dc9d0490bb5f5f1375932b

MarcoFalke commented on src/outputtype.h:
```diff
@@ -19,13 +19,15 @@ enum class OutputType {
     OutputType::P2SH_SEGWIT,
     OutputType::BECH32,
     OutputType::BECH32M,
+    OutputType::UNKNOWN,
```
See https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=50329
